app/models/user_password.rb

```diff
@@ -1,19 +1,55 @@
 # frozen_string_literal: true
 
 class UserPassword < ActiveRecord::Base
-  validates :user_id, presence: true
+  attr_accessor :password # required to build an instance of this model with password attribute without backing column
+  attr_accessor :salt # TODO: Deprecate once we drop User.salt, this is only for passing through the randomized salt from User
+
+  TARGET_PASSWORD_ALGORITHM =
+    "$pbkdf2-#{Rails.configuration.pbkdf2_algorithm}$i=#{Rails.configuration.pbkdf2_iterations},l=32$"
+  PASSWORD_SALT_LENGTH = 16
+  MAX_PASSWORD_LENGTH = 200
+
+  # validates :user_id, presence: true
 
   validates :user_id,
             uniqueness: {
               scope: :password_expired_at,
             },
             if: -> { password_expired_at.nil? }
 
-  validates :password_hash, presence: true, length: { is: 64 }, uniqueness: { scope: :user_id }
-  validates :password_salt, presence: true, length: { is: 32 }
-  validates :password_algorithm, presence: true, length: { maximum: 64 }
+  # validates :password_hash, presence: true, length: { is: 64 }, uniqueness: { scope: :user_id }
+  # validates :password_salt, presence: true, length: { is: 32 }
+  # validates :password_algorithm, presence: true, length: { maximum: 64 }
+  validate :password_validator
+
+  before_save :ensure_password_is_hashed
```

Comment:
I have never used it yet, but I wonder if this is not a good use case for the new: https://edgeapi.rubyonrails.org/classes/ActiveRecord/Normalization.html